ecs_deploy/ecs.py

```diff
@@ -473,6 +595,267 @@ def apply_container_secrets(self, container, new_secrets, exclusive=False):
+            container=container[u'name'],
+            field=u'systemControls',
+            value=merged,
+            old_value=old_system_controls
+        )
+        self._diff.append(diff)
+
+        container[u'systemControls'] = [
+            {name: value for name,value in e.items()} for e in merged
+        ]
+
+    def set_ulimits(self, ulimits_list, exclusive=False):
+        ulimits = {}
+        for ulimit in ulimits_list:
+            ulimits.setdefault(ulimit[0], [])
+
+            mapping = {}
+            mapping["name"] = ulimit[1]
+            mapping["softLimit"] = int(ulimit[2])
+            mapping["hardLimit"] = int(ulimit[3])
+            ulimits[ulimit[0]].append(mapping)
+            
+        self.validate_container_options(**ulimits)
+        for container in self.containers:
+            if container[u'name'] in ulimits:
+                self.apply_container_ulimits(
+                    container=container,
+                    new_ulimits=ulimits[container[u'name']],
+                    exclusive=exclusive,
+                )
+            elif exclusive is True:
+                self.apply_container_ulimits(
+                    container=container,
+                    new_ulimits={},
+                    exclusive=exclusive,
+                )
+
+    def apply_container_ulimits(self, container, new_ulimits, exclusive=False):
+        ulimits = container.get('ulimits', [])
+        old_ulimits = [ {name: value for name, value in ulimit.items()} for ulimit in ulimits]
+
+        if exclusive is True:
+            merged = new_ulimits if new_ulimits else []
+        else:
+            merged = copy.deepcopy(old_ulimits)
+            old_ulimit_names = [ulimit["name"] for ulimit in merged]
+            new_ulimit_names = [ulimit["name"] for ulimit in new_ulimits]
+            new = set(new_ulimit_names) - set(old_ulimit_names)
+            new_merged = list()
+            # update
+            for ulimit in merged:
+                for new_ulimit in new_ulimits:
+                    if ulimit["name"] == new_ulimit["name"]:
+                        ulimit.update(new_ulimit)
+            # new
+            for name in new:
+                for new_ulimit in new_ulimits:
+                    if name == new_ulimit["name"]:
+                        new_merged.append(new_ulimit)
+
+            merged.extend(new_merged)
+
+        if old_ulimits == merged:
+            return
+
+        diff = EcsTaskDefinitionDiff(
+            container=container[u'name'],
+            field=u'ulimits',
+            value=merged,
+            old_value=old_ulimits
+        )
+        self._diff.append(diff)
+
+        container[u'ulimits'] = [
+            {name: value for name,value in e.items()} for e in merged
+        ]
+
+    def set_port_mappings(self, port_mappings_list, exclusive=False):
+        port_mappings = {}
+
+        for port_mapping in port_mappings_list:
+            port_mappings.setdefault(port_mapping[0], [])
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

You are using the `dict.setdefault()` method. We recommend using the `defaultdict` object to handle missing items in internal state because the default value is automatically generated when you access or modify a missing key. `defaultdict` can also be easier to use and understand.

[Learn more](https://docs.python.org/3/library/collections.html#collections.defaultdict)

Reply:
Should be fine now.